cmd/monitoring/monitoring.go

```diff
@@ -0,0 +1,219 @@
+	interval  = flag.Duration("interval", 100*time.Millisecond, "check interval with dimension. Example: 1000ms ")
+	logerrors = flag.Bool("logerrors", false, "log initial errors")
+	outputdir = flag.String("outputdir", "./", "output directory")
+)
+
+type monitor struct {
+	log    *logrus.Entry
+	pipcli azureclient.PublicIPAddressesClient
+
+	resourceGroup  string
+	subscriptionID string
+
+	instances []instance
+}
+
+type instance struct {
+	hostname string
+	b        *blackbox.Config
+}
+
+func (m *monitor) init(ctx context.Context, log *logrus.Entry) error {
+	authorizer, err := azureclient.NewAuthorizerFromEnvironment("")
+	if err != nil {
+		return err
+	}
+	if os.Getenv("RESOURCEGROUP") == "" {
+		return fmt.Errorf("RESOURCEGROUP environment variable must be set")
+	}
+	m.resourceGroup = os.Getenv("RESOURCEGROUP")
+	m.subscriptionID = os.Getenv("AZURE_SUBSCRIPTION_ID")
+	m.pipcli = azureclient.NewPublicIPAddressesClient(ctx, m.subscriptionID, authorizer)
+	m.log = log
+
+	return nil
+}
+
+func (m *monitor) listResourceGroupMonitoringHostnames(ctx context.Context, subscriptionID, resourceGroup string) (hostnames []string, err error) {
+	// get all external IP's used by VMSS
+	for {
+		hostnames = []string{}
+		for iter, err := m.pipcli.ListVirtualMachineScaleSetPublicIPAddressesComplete(ctx, resourceGroup, "ss-master"); iter.NotDone(); err = iter.Next() {
+			if err != nil {
+				m.log.Debug("waiting for url")
+				time.Sleep(5 * time.Second)
+			} else if iter.Value().IPAddress != nil {
+				hostnames = append(hostnames, *iter.Value().IPAddress)
+			}
+		}
+		if err == nil && len(hostnames) == 3 {
+			break
+		}
+	}
+	// get api server hostname
+	for {
+		ip, err := m.pipcli.Get(ctx, resourceGroup, "ip-apiserver", "")
+		if err != nil {
+			m.log.Debug("waiting for url")
+			time.Sleep(5 * time.Second)
+		} else if err == nil && ip.Location != nil {
+			hostnames = append(hostnames, fmt.Sprintf("%s.%s.cloudapp.azure.com", *ip.DNSSettings.DomainNameLabel, *ip.Location))
+			break
+		}
+	}
+	// get dedicated routes we want to monitor
+	// these routes will not be available at creation time, so we will not check for their availability
+	oc, err := loadOCConfig()
+	if err != nil {
+		m.log.Warn("failed to load OpenShiftManagedCluster config, will not monitor routes")
+	} else {
+		hostnames = append(hostnames, fmt.Sprintf("canary-openshift-azure-monitoring.%s", oc.Properties.RouterProfiles[0].PublicSubdomain))
+	}
+	return hostnames, nil
+}
+
+func (m *monitor) run(ctx context.Context) error {
+	ch := make(chan os.Signal)
+	signal.Notify(ch, os.Interrupt, syscall.SIGTERM)
+
+	// we will close this guard when we load all monitors and clean-up is needed
+	bootCtx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+	go func(ctx context.Context) {
+		for {
+			select {
+			case <-ctx.Done():
+				return
+			case <-ch:
+				os.Exit(0)
+			}
+		}
+	}(bootCtx)
```

Comment:
@jim-minter added this for cleaning up the script in the initial state.
It can be done with only this go routine by passing `m.instances` as an address into goroutine for cleanups or making a channel with data for goroutine to consume. But this makes it simple for now. fix as follow-up? so we can start testing?